v0100/srclib/dpstub.asm

```diff
@@ -29,6 +29,14 @@ __start:
         mov     [psp_seg], es
         mov     [cs_seg], cs
 
+        ; release unused DOS memory
+        mov     sp, stack_top
+        mov     bx, end_of_memory
+        shr     bx, 4
+        add     bx, 0x10   ; add psp size
```

Comment:
bx is incorrectly calculated. The memory contains PSP info, code segment, data/stack segment, in that order. All three parts are allocated as one block that begins with the PSP.
The formula should be something like `(ss-es) + (end_of_memory shr 4)`.

Reply:
I don't think I understand this.
I now do this:
`mov     bx, end_of_memory - __start + 0x100`
I made sure that __start starts
immediately after PSP, so I think
that should be correct now.
What do you think?

Reply:
end_of_memory and __start are in different sections (.bss and .text, respectively).
Those sections are not combined into one section or segment.
They end up being in two different segments per the _small memory model_, in which code resides in its own segment and data/stack resides in its own different segment.

Therefore you cannot simply subtract offsets in different sections from one another.

You want to change the size of the block that begins with the PSP.
Its start is ES:0.
Its end is SS:FFFFH or smth like it.

SS-ES is gives you the number of 16-byte paragraphs occupied by the PSP and code.
end_of_memory shr 4 gives you the number of 16-byte paragraphs occupied by .data, .bss and the stack.
You add these together, this is your new/shrinked PSP segment size in paragraphs.

Reply:
Done, so ok now?